src/NosCore.Controllers/ExchangePacketController.cs

```diff
@@ -0,0 +1,251 @@
+using Serilog;
+
+namespace NosCore.Controllers
+{
+    public class ExchangePacketController : PacketController
+    {
+        private readonly WorldConfiguration _worldConfiguration;
+        private readonly ExchangeAccessService _exchangeAccessService;
+        private readonly ILogger _logger = Logger.GetLoggerConfiguration().CreateLogger();
+        
+        public ExchangePacketController(WorldConfiguration worldConfiguration, ExchangeAccessService exchangeAccessService)
+        {
+            _worldConfiguration = worldConfiguration;
+            _exchangeAccessService = exchangeAccessService;
+        }
+
+        [UsedImplicitly]
+        public ExchangePacketController()
+        {
+
+        }
+
+        [UsedImplicitly]
+        public void ExchangeList(ExcListPacket packet)
+        {
+            if (packet.Gold > Session.Character.Gold || packet.BankGold > Session.Account.BankMoney)
+            {
+                return;
+            }
+
+            var list = new ServerExcListPacket
+            {
+                SenderType = SenderType.Server,
+                VisualId = Session.Character.VisualId,
+                Gold = packet.Gold,
+                BankGold = packet.BankGold,
+                SubPackets = new List<ServerExcListSubPacket>()
+            };
+            
+            var target = Broadcaster.Instance.GetCharacter(s => s.VisualId == Session.Character.ExchangeData.TargetVisualId && s.MapInstanceId == Session.Character.MapInstanceId) as Character;
+
+
+            if (packet.SubPackets.Any())
+            {
+                byte i = 0;
+                foreach (var value in packet.SubPackets)
+                {
+                    if (value.PocketType == PocketType.Bazaar)
+                    {
+                        return;
+                    }
+
+                    var item = Session.Character.Inventory.LoadBySlotAndType<IItemInstance>(value.Slot, value.PocketType);
+
+                    if (item == null || item.Amount <= 0 || item.Amount < value.Amount)
+                    {
+                        return;
+                    }
+
+                    if (!item.Item.IsTradable)
+                    {
+                        _exchangeAccessService.CloseExchange(Session, target?.Session, ExchangeCloseType.Failure);
+                        return;
+                    }
+
+                    var itemCpy = (IItemInstance)item.Clone();
+                    itemCpy.Amount = value.Amount;
+                    Session.Character.ExchangeData.ExchangeItems.TryAdd(i, itemCpy);
+
+                    var subPacket = new ServerExcListSubPacket
+                    {
+                        ExchangeSlot = i,
+                        PocketType = value.PocketType,
+                        ItemVnum = itemCpy.ItemVNum
+                    };
+
+                    if (value.PocketType == PocketType.Equipment)
+                    {
+                        subPacket.Amount = itemCpy.Rare;
```

Comment:
AmountOrRare also move this into the constructor. Upgrade can be shared it will be 0 just add a ternary for rare